### PR TITLE
Improve VisualScript UX

### PR DESCRIPTION
**Pull request by fire**
- [x] Prototype faster function call ux.
- [x] Work on general search ux.
- [x] Able to create nodes from search.
- [x] Show class for variables but not methods.
- [x] Get actions search working.
- [x] Descriptions now show for both methods and properties.
- [x] Enable zooming on mouse wheel up and down. Removed due to bugginess.
- [x] Make the drag trigger on right mouse button.
- [x] Search now shows for action visual script nodes.

### CODE REVIEW

**Comment by reduz** [comment]
I think from all the feedback I read, a main change that should be done in VisualScript (even if it means breaking compatibility), is getting rid of per-function graphs and just make a single canvas graph for all functions. A function would be a single entry point into it.

And from there, allowing to create custom nodes from existing sections of node layouts in the graph for reusing or simplification of the graph.

The current way I think is not very "discoverable" by users.

**Comment by fire** [comment]
Currently there's a bug in this pull request with int / real types not activating search.

**Comment by fire** [comment]
After investigation this is a bug in the original visual script. It's in 3.0.2 stable.

**Comment by fire** [comment]
Does anyone know how this is constructed?

help

**Comment by djrm** [comment]
as far as i can tell mouse wheel zoom was rejected because the zoom quality is really bad.

**Comment by akien-mga** [comment]
Side note: please keep in mind that each time you push a new commit, you trigger a new CI build on Travis and AppVeyor:
screenshot_20180517_084539
screenshot_20180517_084618

Sure we don't pay for this processing power ourselves, but it means that other PRs/commits may be stalled due to the long build queue of already-obsolete commits. So if you need to iterate very quickly with new commits (which is totally fine), might be worth holding off on `git push` a bit while developing, or push to another branch and update the PR once ready to merge/review.

**Comment by fire** [comment]
Master is broken with the visual script dock. Patching with this gives a broken dock. This weekend I'm trying to do a game jam so there won't be updates.

**Comment by Calinou** [comment]
> So if you need to iterate very quickly with new commits (which is totally fine), might be worth holding off on git push a bit while developing, or push to another branch and update the PR once ready to merge/review.

Note that you can also add `[ci skip]` in the title of commit messages, as explained in the [Travis CI documentation](https://docs.travis-ci.com/user/customizing-the-build/#Skipping-a-build). It has to be in the title, otherwise [AppVeyor won't skip the build](https://github.com/appveyor/ci/issues/1865). Of course, you should remove it when ready for review using `git commit --amend`.

**Comment by fire** [comment]
The current code is now rebased on top of master.

I experimented having right click (couldn't get the drag to work) open a search window for visual script nodes. What do you think? I had problems getting the position of the mouse before the window is open. Getting the mouse position is straightforward.

**Comment by fire** [comment]
Summary of irc chat with Zylann:

1. new select_from_* look very similar

\<iFire> It's basically a general select. I tried to remove the existing functions but ran into problems. thought about either calling the existing functions or duplicating the code. duplicating the code felt more straightforward

\<Zylann> iFire also you are mixing properties and functions by doing this? I dunno, I personally don't like duplication unless meaning is completely different

\<iFire> the idea that when I pull from a Vector2 node I search for all the possible visual nodes, which include setting a vec2 or calling functions that have a vec2 input

\<iFire> the first example is a property and the second is a function. This is the "mixing properties and functions".

\<Zylann> iFire I don't know, maybe there can be as many functions, but some of their contents is needlessly copied. Such as popup_centered_ratio(0.6); That doesn't need to be different. So you can put it in a common place maybe. Basically anything that doesn't *need* to be different can be in a common place

2. Have commented code in the graph edit code

\<Zylann> Some of which is just the same code \<iFire> Not sure how to remove the first duplication

3. <Zylann> iFire: why did you add RMB for panning? 

It A) wasn't being used B) it is used in the same way in another engine's visual script C) a related feature is RMB for doing search while !dragging

\<Zylann> A) isn't a valid argument alone IMO^^
\<iFire> It felt natural to use RMB for dragging is my reasoning, but I'm not the only game engine that thinks that way.
\<Zylann> I'd never use RMB for panning, I never used any software doing that. Maybe visual script editors like to feel special
\<iFire> LMB is used for dragging nodes and sequence lines. middle button is usable for dragging but is not intuitive and it also already implemented
\<Zylann> Not intuitive? There is so many software using it. Either alone or with Shift
\<iFire> Basically I added another way to 2 existing ways to drag.
\<Zylann> Only problem is trackpads, which use spacebar I think
\<iFire> So it would be RMB to drag, MMB to drag, I think shift to drag (don't remember)
\<Zylann> Yeah but it feels weird to add a third way to pan while I think nobody asked for it?
\<iFire> Zylann: I want to poll other people, but it was a major pain point for me. Like hand out demo of the visual script changes.
\<Zylann> Also RMB needs care because we may use it for context menus. What is usually done with click+drag things  (be it RMB or LMB) is that they include a small margin before dragging actually starts. Because if not, you may trigger dragging just because your mouse moved by 1px. \<Zylann> While you might have intended to select or open a context menu. If that is done, then you can use RMB and LMB for dragging too. Otherwise it's a bit more risky
\<iFire> My actually goal is for RMB to single click open a search window for visual script nodes, if RMB DRAG then pan.
\<Zylann> But that margin thing wouldn't need to be done on per case basis, UI framework usually do that in backend
\<Zylann> iFire it's just a UX concern among others
\<iFire> There's a github issue about drag not integrated #19457
\<Zylann> iFire my main concern with RMB is that now we have 3 ways of doing the same thing^^. It's not necessarily bad
\<iFire> we should have more ways imho
\<Zylann> But more ways = more cases to maintain
\<iFire> Anyways in my pull request there is RMB panning but no RMB DRAG causing a search window. I had problems storing the initital point where the RMB was in the prototype the RMB loc is where I clicked when I exited the search window.

Note: https://github.com/godotengine/godot/issues/18831 Design documentation.

**Comment by fire** [comment]
There is a bug with my patch of visual script where calling self and then setting a variable has an extraneous sequence line. I'll have to investigate.

**Comment by fire** [comment]
Fixed the bug with extra sequence line.

**Comment by MarianoGnu** [comment]
> I think from all the feedback I read, a main change that should be done in VisualScript (even if it means breaking compatibility), is getting rid of per-function graphs and just make a single canvas graph for all functions. A function would be a single entry point into it.

I just dont understand why users wants this, how could  bigass masive graph be easier to read and interpreat than smaller ones :/

**Comment by fire** [comment]
One advantage of a big graph is that seeing the logic flow of multiple nodes is easier at a glance. Moving them is also easier.

**Comment by abaky** [comment]
I don't know if this has been discussed here or anywhere else, but would it be possible to show in play mode on the line that connects logic blocks what kind of a value is being pushed to the next node? For reference I would point to [Unity Bolt add on](https://ludiq.io/bolt). It makes it very fast and easy to see values that are being pushed around. It also has a nice and clean UI, overall very intuitive tool (I have no affiliation with Bolt, just find it to be a very nice solution and think that a lot of inspiration could be drawn from it).

**Comment by fire** [comment]
@abaky That sounds like a feature to debug Visual Script.

One has to look at the debugging api to see if it has the info required without using the in-editor playback feature. 
https://github.com/godotengine/godot/pull/13354

File: scene/gui/graph_edit.cpp
Comment (reduz) [inline scene/gui/graph_edit.cpp]: No dependencies like this should happen, I suggest you find a way to extend GraphEdit to be useful for what you need.

File: scene/gui/graph_edit.cpp
Comment (reduz) [inline scene/gui/graph_edit.cpp]: No dependencies like this should happen, I suggest you find a way to extend GraphEdit to be useful for what you need.

**Review by reduz — CHANGES_REQUESTED** [review]

File: editor/property_selector.cpp
Comment (akien-mga) [inline editor/property_selector.cpp]: `editor/` files can't be made to depend on a module, as said module could be disabled and compilation will fail. If you need to make VisualScript editor features, they should go in `modules/visual_script/`.

**Review by akien-mga — DISMISSED** [review]

**Comment by akien-mga** [comment]
Beyond the changes @reduz and I requested, it would be nice to have a TL;DR of what the main changes are, ideally with some screenshots/gifs, to make it easier to understand what's changing.

File: editor/property_selector.cpp
Comment (fire) [reply editor/property_selector.cpp]: Resolved this issue by moving property selector to modules/visual_script/.

File: scene/gui/graph_edit.cpp
Comment (fire) [reply scene/gui/graph_edit.cpp]: Resolved this issue by turning the dependency to a signal call.

File: scene/gui/graph_edit.cpp
Comment (fire) [reply scene/gui/graph_edit.cpp]: Resolved this issue by turning the dependency to a signal call.

**Comment by fire** [comment]
TL;DR

https://imgur.com/a/S9bjaC4

* Prototype faster function call ux.
    Instead of dragging on the node / sequences calling a 3 element menu, directly do the search. 
* Type is autocompleted.
* Able to create nodes from search.
* Operators, builtins and constructors show type name. Fix several problems with port connections.
* Add icons to search items.
* In shift-a mode search everything.
* For the drop zones, make all the rect2 areas the same size.
* Add bigger hit zones to node connect.
* Add vs constructors in shift-a menu.

**Comment by fire** [comment]
As a zip.

[videos.zip](https://github.com/godotengine/godot/files/2206749/videos.zip)

File: editor/editor_properties.cpp
Comment (akien-mga) [inline editor/editor_properties.cpp]: I wouldn't put the default argument as a comment inline like that, it looks weird. But actually `p_category` isn't used anywhere, so it does not need to be added.

File: modules/visual_script/visual_script_property_selector.cpp
Comment (akien-mga) [inline modules/visual_script/visual_script_property_selector.cpp]: The closing `*/` is misaligned.

File: modules/visual_script/visual_script_property_selector.h
Comment (akien-mga) [inline modules/visual_script/visual_script_property_selector.h]: It should be `visual_script_property_selector.h`.

File: scene/gui/graph_edit.cpp
Comment (akien-mga) [inline scene/gui/graph_edit.cpp]: This doesn't seem to be called anywhere?

Comment (akien-mga) [reply scene/gui/graph_edit.cpp]: You do connect to the signal in `modules/visual_script/visual_script_editor.cpp`, but as far as I can tell it will never be emitted.

File: modules/visual_script/visual_script_property_selector.h
Comment (fire) [reply modules/visual_script/visual_script_property_selector.h]: Fixed.

File: modules/visual_script/visual_script_property_selector.cpp
Comment (fire) [reply modules/visual_script/visual_script_property_selector.cpp]: Fixed.

File: editor/editor_properties.cpp
Comment (fire) [reply editor/editor_properties.cpp]: Is removed.

File: scene/gui/graph_edit.cpp
Comment (reduz) [inline scene/gui/graph_edit.cpp]: This is pretty hacky, I wonder if there can be a better way to do this..

File: scene/gui/graph_edit.cpp
Comment (reduz) [inline scene/gui/graph_edit.cpp]: I don t think anywhere in Godot you do this with RMB, so adding it here is probably not a good idea

File: scene/gui/graph_edit.cpp
Comment (reduz) [inline scene/gui/graph_edit.cpp]: you dont need to do this here, you can conenct to the input_event signal of graph_edit,  intercept this key and accept the event

**Review by reduz — CHANGES_REQUESTED** [review]

File: scene/gui/graph_edit.cpp
Comment (fire) [reply scene/gui/graph_edit.cpp]: What do you mean?

File: scene/gui/graph_edit.cpp
Comment (fire) [reply scene/gui/graph_edit.cpp]: Done,  uploading revision.

File: scene/gui/graph_edit.cpp
Comment (fire) [reply scene/gui/graph_edit.cpp]: I extracted the common code into create_hot_zones()

File: scene/gui/graph_edit.cpp
Comment (fire) [reply scene/gui/graph_edit.cpp]: Developer feedback is that RMB should drag. 

\<BrahRah>The RMB seems to have 0 functionality atm. there should be at least a pop up menu. The mouse plays a huge role in Visual Coding and thus needs to be implemented really really well. - https://github.com/godotengine/godot/issues/19862

File: scene/gui/graph_edit.cpp
Comment (fire) [reply scene/gui/graph_edit.cpp]: Moved constants into default theme.

File: scene/resources/default_theme/default_theme.cpp
Comment (reduz) [inline scene/resources/default_theme/default_theme.cpp]: why cant this be pixels, like everything else?

File: scene/gui/graph_edit.cpp
Comment (akien-mga) [reply scene/gui/graph_edit.cpp]: I agree with @reduz and @vnen, it's inconsistent with other editor plugins and the functionality is already available with MMB and LMB + Space.

RMB should be kept for another meaningful operation.

File: scene/gui/graph_edit.cpp
Comment (reduz) [inline scene/gui/graph_edit.cpp]: this is too messy, just put a radius in pixels as a constant in default_theme and dont use the port texture size

**Review by reduz — CHANGES_REQUESTED** [review]

File: scene/resources/default_theme/default_theme.cpp
Comment (fire) [reply scene/resources/default_theme/default_theme.cpp]: Done. Moved the two grab_r variables into the constant.

File: scene/gui/graph_edit.cpp
Comment (fire) [reply scene/gui/graph_edit.cpp]: Done. Moved the two grab_r variables into the constant. Ditto.

**Comment by aaronfranke** [comment]
In the future, I recommend using GitHub checklists instead of crossing out list items.

- [x] Checked item
- [] Unchecked item

You can make these by prefixing each line with `- [x]` or `- []`

**Pull request merged**
